src/Workspaces/Core/Portable/Shared/Utilities/BloomFilter.cs

```diff
@@ -395,4 +398,73 @@ private static bool IsEquivalent(BitArray array1, BitArray array2)
 
         return true;
     }
+
+    /// <summary>
+    /// Provides mechanism to efficiently obtain bloom filter hash for a value. Backed by a single element cache.
+    /// </summary>
+    internal class BloomFilterHash
+    {
+        private static BloomFilterHash? s_cachedHash;
+
+        private readonly string _value;
+        private readonly bool _isCaseSensitive;
+        private readonly ImmutableArray<int> _hashes;
+
+        private BloomFilterHash(string value, BloomFilter filter)
+        {
+            _value = value;
+            _isCaseSensitive = filter._isCaseSensitive;
+
+            var hashBuilder = new FixedSizeArrayBuilder<int>(filter._hashFunctionCount);
+
+            for (var i = 0; i < filter._hashFunctionCount; i++)
+                hashBuilder.Add(filter.GetBitArrayIndex(value, i));
+
+            _hashes = hashBuilder.MoveToImmutable();
+        }
+
+        /// <summary>
+        /// Although calculating this hash isn't terribly expensive, it does involve multiple
+        /// (usually around 13) hashings of the string (the actual count is <see cref="BloomFilter._hashFunctionCount"/>).
+        /// The typical usage pattern of bloom filters is that some operation (eg: find references)
+        /// requires asking a multitude of bloom filters whether a particular value is likely contained.
+        /// The vast majority of those bloom filters will end up hashing that string to the same values, so
+        /// we put those values into a simple cache and see if it can be used before calculating.
+        /// Local testing has put the hit rate of this at around 99%.
+        /// </summary>
+        public static ImmutableArray<int> GetOrCreateHashArray(string value, BloomFilter filter)
+        {
+            var cachedHash = s_cachedHash;
+
+            if (cachedHash == null
+                || cachedHash._isCaseSensitive != filter._isCaseSensitive
+                || cachedHash._hashes.Length < filter._hashFunctionCount
```

Comment:
Why is this a `<` and not a `!=`?

Reply:
Because a longer hash of arrays is fine. The caller may not use all the values in the array we hand back, but that's not a big deal.